src/net/tls.cc

```diff
@@ -1217,7 +1223,11 @@ class session : public enable_lw_shared_from_this<session> {
             gnutls_x509_crt_deinit(peer_leaf_cert);
 
             if (ec || ec2) {
-                throw std::runtime_error("error while extracting certificate DN strings");
+                std::runtime_error err("error while extracting certificate DN strings");
+                if (fetch_dn) {
+                    _session_dn->set_exception(err);
```

Comment:
if an exception is thrown, do you think it will be likely that the caller won't ever call `get_distinguished_name` because they'll experience as error before they get their hands on a healthy connected socket?

one downside to `set_exception` here is that if the caller (ie redpanda) _doesn't_ consume the now-exceptional-future then seastar will log errors complaining about "abandoned exceptional future".

I think there are a few ways to handle this. one way might be to set the exception, but in the destructor throw away the exception in the promise if it exists. another option might be to record an error, but only manifest the exceptional future if `get_distinguished_name` is called.

Reply:
Another approach that I'm going to take is to get rid of `set_exception` uses. The exception thrown from `verify` gets propagated to the code that uses the socket anyway. We don't need to duplicate it to pass through the future returned by `get_distinguished_name`. In case of failure the caller will get an exception and the future will return `std::nullopt` instead of exception.